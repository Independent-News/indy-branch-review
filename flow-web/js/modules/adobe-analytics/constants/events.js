export const INTERNAL_SEARCH = 'event1';
export const TOPIC_PAGE_LOAD = 'event2';
export const GALLERY_VIEW = 'event3';
export const AUTO_REFRESH = 'event6';
export const ARTICLE_VIEW = 'event7';
export const SELLING_PAGE_LOADED = 'event8';
export const SIGN_IN = 'event9';
export const ARTICLE_SHARE = 'event10';
export const COMMENT_ADD = 'event12';
export const READ_MORE_CLICK = 'event13';
export const AUTO_GALLERY_VIEW_COUNT = 'event14';
export const PREMIUM_COMMENT_ADD = 'event16';
export const ARTICLE_BOOKMARKED = 'event17';
export const RHR_PREMIUM_COMPONENT_CLICK = 'event18';
export const RHR_MOST_POPULAR_CLICK = 'event19';
export const RHR_INDY100_TRENDING_CLICK = 'event20';
export const SUBSCRIBE_BUTTON_CLICK = 'event21';
export const REG_PAGE_LOAD = 'event22';
export const REG_SUCCESS = 'event23';
export const MARKETING_OPT_IN_CLICK = 'event24';
export const PAYMENT_DETAILS_LOADED = 'event25';
export const PROMO_CODE_APPLIED = 'event26';
export const PAYMENT_DETAILS_SUBMITTED = 'event27';
export const SUBSCRIPTION_SUCCESS = 'event28';
export const AUTO_RENEW_DISABLED = 'event29';
export const RHR_POPULAR_VIDEOS_CLICK = 'event30';
export const RHR_SPONSORED_FEATURES_CLICK = 'event31';
export const PAYMENT_UNSUCCESSFUL = 'event32';
export const DONATION_BUTTON_CLICK = 'event34';
export const TAX_CONFIRMATION_PROMPT = 'event35';
export const DONATION_ARTICLE_BOTTOM_CLICK = 'event36';
export const NAV_CLICK = 'event37';
export const ARTICLE_INLINE_LINK_CLICK = 'event39';
export const ARTICLE_TOPIC_TAG_CLICK = 'event40';
export const ARTICLE_SHARE_ICON_CLICK = 'event41';
export const COMMENTS_SHOW_CLICK = 'event42';
export const COMMENTS_TAB_CLICK = 'event43';
export const NOTIFICATION_PROMPT_LOAD = 'event44';
export const NOTIFICATION_PROMPT_ACCEPT = 'event45';
export const NOTIFICATION_PROMPT_DISMISS = 'event46';
export const HOMEPAGE_LINK_CLICK = 'event46'; // this is used, but Graham isn't sure what for
export const MOST_WATCHED_CLICK = 'event51';
export const COMMENTS_LOAD_MORE = 'event53';
export const RELATED_ARTICLE_BULLET_CLICK = 'event57';
export const LIVE_BLOG_INLINE_ARTICLE_CLICK = 'event58';
export const LIVE_BLOG_KEY_POINTS_CLICK = 'event60';
export const REG_GATE_SHOWN = 'event62';
export const NEWSLETTER_LITE_REG_SIGNUP = 'event63';
export const NEWSLETTER_PREFS_SUBSCRIBE = 'event64';
export const NEWSLETTER_PREFS_UNSUBSCRIBE = 'event65';
export const NEWSLETTER_PAGE_SIGNUP_CLICK = 'event66';
export const REGISTER_NOW_CLICK = 'event68';
export const EBOOKS_CLICK = 'event69';
export const PREMIUM_EVENT_CLICK = 'event70';
export const ARTICLE_INLINE_LINK_TOPIC_PAGE = 'event72';
export const ARTICLE_INLINE_LINK_NON_TOPIC_PAGE = 'event73';
export const SUBSCRIPTION_FORM_FIELD_FOCUS = 'event74';
export const SUBSCRIPTION_METHOD_SELECTED = 'event75';
export const SUBSCRIPTION_PACKAGE_CHANGED = 'event76';
export const MOST_COMMENTED_SECTION_CLICK = 'event78';
export const AFFILIATE_LINK_CLICK = 'event79';
export const BURGER_MENU_CLICK = 'event80';
export const EDITION_LINK_CLICK = 'event81';
export const VOUCHER_TICKER_LINK_CLICK = 'event82';
export const VOUCHER_TICKER_ARROW_CLICK = 'event83';
export const PIANO_EXPERIENCE_IMPRESSION = 'event84';
export const ACCOUNT_OVERVIEW_CLICK = 'event85';
export const CANCEL_JOURNEY_CLICK = 'event86';
export const CANCELLATION_JOURNEY_STAGE = 'event87';
export const WEBPUSH_SIGNUP_CLICK = 'event106';

export const BUY_NOW_CLICK = 'event93';
export const PRICE_COMPARISON_BUY_NOW_CLICK = 'event107';
export const PRODUCT_CAROUSEL_LOADED = 'event97';
export const PRODUCT_CAROUSEL_BUY_NOW_CLICK = 'event108';
export const PRODUCT_CAROUSEL_READ_REVIEW_CLICK = 'event99';
export const PRODUCT_COMPARISON_CHECKBOX_SELECTED = 'event100';
export const PRODUCT_COMPARISON_MODAL_OPEN = 'event101';

export const ADBLOCK_DETECTED = 'event105';

export const SUPPORT_NSC_LINK_CLICK = 'event110';
export const DONATE_PAGE_LOAD = 'event110';
