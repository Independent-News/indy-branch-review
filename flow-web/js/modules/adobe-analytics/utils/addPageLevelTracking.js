/* globals s_object */

export default ({ props = {}, events = [] }) => {
  events.forEach((evt) => {
    s_object.events = s_object.apl(s_object.events, evt, ',', ',', false);
  });

  Object.entries(props).forEach(([key, value]) => {
    s_object[key] = value;
  });
};
