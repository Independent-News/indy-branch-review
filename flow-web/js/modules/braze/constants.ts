export const BRAZE_EVENT_AUTORENEWAL_ON = 'turned_auto_renewalON';
export const BRAZE_EVENT_AUTORENEWAL_OFF = 'turned_auto_renewalOFF';
export const BRAZE_EVENT_PREMIUM_METER_EXPIRED = 'premium_meter_expired';
export const BRAZE_EVENT_REGISTRATION_COMPLETE = 'registration_complete';
export const BRAZE_EVENT_SUBSCRIPTION_START = 'subscription_start';
export const BRAZE_EVENT_SUBSCRIPTION_FINISH = 'subscription_finish';
export const BRAZE_EVENT_SUBSCRIPTION_ONBOARDING_SURVEY =
  'subscription_onboarding_survey';
export const BRAZE_EVENT_DONATION_START = 'donation_start';
export const BRAZE_EVENT_DONATION_FINISH = 'donation_finish';
