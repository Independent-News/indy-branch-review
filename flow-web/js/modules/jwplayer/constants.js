export const PARAM_AD_POS = 'ap';
