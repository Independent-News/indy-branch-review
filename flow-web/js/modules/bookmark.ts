import { COOKIE_LOGGED_IN } from '#app/constants/cookies';

import {
  CLASS_BOOKMARK_BUTTON,
  CLASS_BOOKMARK_NOTIFICATION,
  CLASS_BOOKMARK_NOTIFICATION_CLOSE,
  CLASS_BOOKMARK_NOTIFICATION_SHOW,
  CLASS_BOOKMARK_NOTIFICATION_LOGIN,
} from '#app/constants/classNames';
import { BOOKMARK_OVERLAY_LOGIN_RSM } from '#app/constants/regSourceMethods';

import { InternalApi } from '../utils/internalApi';

import * as thisModule from './bookmark';
import { getCookie } from './cookie';
import { dispatchArticleBookmark } from './custom-event-dispatchers';
import { openLoginDrawer } from './loginDrawer';

const initNotification = () => {
  const notification = document.querySelector(
    `.${CLASS_BOOKMARK_NOTIFICATION}`,
  );
  if (!notification) {
    return;
  }

  const closeNotification = notification.querySelector(
    `.${CLASS_BOOKMARK_NOTIFICATION_CLOSE}`,
  );
  if (closeNotification) {
    closeNotification.addEventListener('click', () => {
      notification.classList.remove(CLASS_BOOKMARK_NOTIFICATION_SHOW);
    });
  }

  const ignoreCheckbox = notification.querySelector('input');
  if (ignoreCheckbox) {
    ignoreCheckbox.addEventListener('click', () => {
      localStorage.setItem('ignore-bookmark', 'true');
    });
  }
};

const initLightbox = () => {
  if (!window.JSGlobals.featureFlags.feat__no_amp) {
    const loginBtns = [
      ...document.querySelectorAll(`.${CLASS_BOOKMARK_NOTIFICATION_LOGIN}`),
    ];
    loginBtns.forEach((btn) => {
      btn.addEventListener('click', () => {
        const closeButton = document.querySelector('#bookmark-lightbox .close');
        if (closeButton instanceof HTMLButtonElement) {
          closeButton.click();
        }

        setTimeout(() => {
          openLoginDrawer(BOOKMARK_OVERLAY_LOGIN_RSM);
        }, 400);
      });
    });
  }
};

const showBookmarkNotification = () => {
  const notification = document.querySelector('.bookmark-notification');
  const ignoreNotification = localStorage.getItem('ignore-bookmark');
  if (!notification || ignoreNotification) {
    return;
  }

  notification.classList.add(CLASS_BOOKMARK_NOTIFICATION_SHOW);
  setTimeout(() => {
    notification.classList.remove(CLASS_BOOKMARK_NOTIFICATION_SHOW);
  }, 3000);
};

const addBookmarkLocalStorage = (articleId: string) => {
  const bookmarks = JSON.parse(
    localStorage.getItem('premiumBookmarks') || '[]',
  );
  bookmarks.push(articleId);
  localStorage.setItem('premiumBookmarks', JSON.stringify(bookmarks));
};

export const removeBookmarkLocalStorage = (articleId: string) => {
  let bookmarks = JSON.parse(localStorage.getItem('premiumBookmarks') || '[]');

  bookmarks = bookmarks.filter((a: string) => a !== articleId);
  localStorage.setItem('premiumBookmarks', JSON.stringify(bookmarks));
};

export const deleteBookmark = async (
  articleId: string,
  button: HTMLButtonElement,
) => {
  if (button.classList.contains('submitting')) {
    return;
  }
  try {
    button.classList.add('submitting');
    const res = await InternalApi.delete(`bookmarks/${articleId}`);
    if (res.ok) {
      [...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`)].forEach(
        (btn) => btn.classList.remove('saved'),
      );
      button.classList.remove('submitting');
      thisModule.removeBookmarkLocalStorage(articleId);
      return;
    }
    throw new Error('Network response was not ok.');
  } catch (error) {
    [...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`)].forEach((btn) =>
      btn.classList.add('saved'),
    );
    button.classList.remove('submitting');
    console.error('Fetch failed: ', error.message);
  }
};

export const submitBookmark = async (
  articleId: string,
  button: HTMLButtonElement,
) => {
  if (button.classList.contains('submitting')) {
    return;
  }
  button.classList.add('submitting');
  const data = { articleId };
  InternalApi.post('bookmarks', data)
    .then((response) => {
      if (response.ok) {
        return response.json();
      }
      throw new Error('Network response was not ok.');
    })
    .then((response) => {
      if (response.saved) {
        [...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`)].forEach(
          (btn) => btn.classList.add('saved'),
        );
        addBookmarkLocalStorage(articleId);
        showBookmarkNotification();
        dispatchArticleBookmark();
      } else {
        [...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`)].forEach(
          (btn) => btn.classList.remove('saved'),
        );
        removeBookmarkLocalStorage(articleId);
      }
      button.classList.remove('submitting');
    })
    .catch((error) => {
      button.classList.remove('submitting');
      console.error('Fetch failed: ', error.message);
    });
};

export const updateBookmarkState = (
  articleId: string,
  button: HTMLButtonElement,
) => {
  const callback = button.classList.contains('saved')
    ? thisModule.deleteBookmark
    : thisModule.submitBookmark;
  callback(articleId, button);
};

const checkSavedBookmark = (articleId: string, button: HTMLButtonElement) => {
  const bookmarks = JSON.parse(
    localStorage.getItem('premiumBookmarks') || '[]',
  );

  if (bookmarks.indexOf(articleId) > -1) {
    button.classList.add('saved');
  } else {
    button.classList.remove('saved');
  }
};

const initBookmark = async () => {
  const bookmarkBtns = [
    ...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`),
  ];
  const bookmarkNotification = document.querySelector(
    `.${CLASS_BOOKMARK_NOTIFICATION}`,
  );

  if (!bookmarkNotification) {
    return;
  }

  bookmarkBtns.forEach((btn) => {
    if (btn instanceof HTMLButtonElement) {
      const articleId = btn.dataset.articleId || '';
      if (getCookie(COOKIE_LOGGED_IN)) {
        checkSavedBookmark(articleId, btn);
      }

      btn.addEventListener('click', () => {
        updateBookmarkState(articleId, btn);
      });
    }
  });
  initLightbox();
  initNotification();
};

export default initBookmark;
