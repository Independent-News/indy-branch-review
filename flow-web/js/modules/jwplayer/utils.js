/* globals JSGlobals */

import { formatMantisRatings } from '#app/util/mantisDataFormatters';

import isAnonymisedCallbackPageLoad from '../../utils/isAnonymisedCallbackPageLoad';
import { AdTagURL } from '../adTag/AdTagURL';
import {
  MACRO_PAGE_URL,
  MACRO_TIMESTAMP,
  MACRO_GDPR,
  MACRO_GDPR_CONSENT,
  MACRO_ADDTL_CONSENT,
} from '../adTag/macros';
import {
  PARAM_SIZE,
  PARAM_AD_UNIT,
  PARAM_REQUEST_MODE,
  PARAM_DFP_REQUEST,
  PARAM_AD_RULE,
  PARAM_ENVIRONMENT,
  PARAM_AD_OUTPUT,
  PARAM_DELAYED_IMPRESSION,
  PARAM_DESCRIPTION_URL,
  PARAM_CORRELATOR,
  PARAM_GDPR,
  PARAM_GDPR_CONSENT,
  PARAM_ADDTL_CONSENT,
} from '../adTag/params';
import {
  getTargeting as apsGetTargeting,
  loadScript as loadApsScript,
} from '../aps';
import { getTargeting as prebidGetTargeting } from '../instreamVideoPrebid';
import { PARAM_AD_POS } from '../jwplayer/constants';
import { log, warn } from '../log';
import {
  getVideoCookieDeprecationLabel,
  getPpid,
} from '../videoPlayer/videoTargeting';

import * as thisModule from './utils';

const isAnonymisedPageLoad = isAnonymisedCallbackPageLoad();

/**
 * Determine if current page is tv hub
 * @returns {boolean}
 */
export const isTV = () => location.pathname.startsWith('/tv/');

/**
 * Get the article Id
 * @returns {string}
 */
export const getArticleId = () => JSGlobals?.pageId?.replace(/\D/g, '') || '';

/**
 * Get the value of "article" in cust_params
 * @returns {string}
 */
export const getCustParamArticle = () =>
  thisModule.isTV() ? 'videohub' : thisModule.getArticleId();

/**
 * Get topic tags string of current page
 * @returns {string}
 */
export const getTopicTags = () => JSGlobals?.topictags?.join() || '';

/**
 * Get the referrer url from document object
 * @returns {string}
 */
export const getReferrer = () => document.referrer.split('?')[0];

/**
 * Get mantis string
 * @returns {string}
 */
export const getMantisTargeting = () =>
  JSGlobals.mantis_channels?.ratings?.length &&
  formatMantisRatings(JSGlobals.mantis_channels.ratings);

/**
 * Get pixels string
 * @returns {string}
 * @todo check if it should be removed
 */
export const getPixelsTargeting = () =>
  JSGlobals.article?.autoVideo
    ? JSGlobals.article.autoVideo === 'video swapped'
      ? 'swapped'
      : 'reset'
    : '';

export const isVerticalMobile = () => {
  const width = window.innerWidth;
  const height = window.innerHeight;
  const isVertical = width / height < 1;
  return isVertical && width < 768;
};

export const isVerticalAd = (adItem) => {
  const width = adItem.ima.ad.getVastMediaWidth();
  const height = adItem.ima.ad.getVastMediaHeight();
  return width / height < 1;
};

export const isPersistentPlayer = () => {
  const topContainer = document.querySelector('.video-top-container.video');
  return !!topContainer?.classList.contains('sticky');
};

export const closeAd = (playerInstance) => {
  const container = playerInstance.getContainer();
  container.classList.remove('fullscreen');

  playerInstance.isFullscreen = false;
  playerInstance.resize('100%', playerInstance.originalHeight);
};

export const fullscreenVideoCheck = (playerInstance) => {
  playerInstance.isFullscreen && closeAd(playerInstance);
};

export const verticalVideoCheck = (playerInstance, adItem) => {
  if (isVerticalMobile() && !isPersistentPlayer() && isVerticalAd(adItem)) {
    const container = playerInstance.getContainer();
    container.classList.add('fullscreen');

    playerInstance.isFullscreen = true;
    playerInstance.originalHeight = playerInstance.getHeight();
    playerInstance.resize('100%', '100%');
  }
};

/**
 * Get the permutive string
 * @returns {string}
 */
export const getPermutive = () =>
  JSON.parse(localStorage._pdfps || '[]')
    .slice(0, 250)
    .join(',');

/**
 * Get adPos from query string
 * @returns {integer|string}
 */
export const getAdPos = () =>
  new URL(window.location).searchParams.get(PARAM_AD_POS) || 0;

/**
 * Get the player state
 * @param {object} playerInstance
 * @returns {string} "float" if it is floating player or "hero"
 */
export const getPlayerState = (playerInstance) =>
  playerInstance.getFloating() ? 'float' : 'hero';

/**
 * Converts a time string in format "HH:MM:ss.mmm" to seconds.
 * @param {string} timeString - The time string to convert.
 * @returns {number} - The integer number of seconds.
 */
export const timeToSeconds = (timeString) => {
  const parts = String(timeString).split(':').map(parseFloat);
  const seconds = parts.pop();
  return Math.floor(parts.reduce((acc, curr) => 60 * curr + acc, seconds));
};

/**
 * Calculates if the given time is greater than or equal to the position, and returns a boolean value.
 * @param {number} position - The position in time to be compared against.
 * @param {number} time - The time to check against the position.
 * @param {Array<number>} elapsed - An array of time values that have already been matched.
 * @returns {boolean} Returns `true` if the position is less than or equal to the time and the time has not been previously matched. Otherwise, returns `false`.
 */
export const timeElapsed = (position, time, elapsed) => {
  const key = time;
  if (elapsed.includes(key)) {
    return false;
  } else {
    if (position >= time) {
      elapsed.push(key);
      return true;
    } else {
      return false;
    }
  }
};

/**
 * Calculates whether a quarter of the time has elapsed for a given position, based on the elapsed time and the total time
 * @param {number} position - The current position
 * @param {number} duration - The duration
 * @param {Array<number>} elapsed - The elapsed time
 * @param {string} [type=''] - An optional string to prefix the console output with
 * @returns {boolean} - True if a quarter of the time has elapsed for the given position, false otherwise
 */
export const quarterTimeElapsed = (position, duration, elapsed, type = '') => {
  const quartile = duration / 4;

  const q1 = thisModule.timeElapsed(position, quartile, elapsed);
  const q2 = thisModule.timeElapsed(position, quartile * 2, elapsed);
  const q3 = thisModule.timeElapsed(position, quartile * 3, elapsed);
  const q4 = thisModule.timeElapsed(position, duration - 0.5, elapsed);

  log(
    `${type} > ${q1 ? 1 : 0} ${q2 ? 1 : 0} ${q3 ? 1 : 0} ${
      q4 ? 1 : 0
    } ${position.toFixed(2)}:${duration}`,
  );

  return q1 || q2 || q3 || q4;
};

/**
 * Generates a unique UUID v4 to assign to the user
 * @returns {string} A string representation of the UUID.
 */
export const generatePermutiveUUID = () => {
  let d = new Date().getTime();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
    const r = (d + Math.random() * 16) % 16 | 0;
    d = Math.floor(d / 16);
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
};

/**
 * Generate client object for permutive tracking
 * @returns {object}
 */
export const getPermutiveClient = () => ({
  url: window.location.href,
  type: 'web',
  domain: window.location.hostname,
  user_agent: navigator.userAgent,
});

export const permutiveEnabled = (consent) =>
  !!(!isAnonymisedPageLoad && consent && window.JSGlobals.permutive);

/**
 * Generate ad object with given ad item for permutive tracking
 * @returns {object}
 */
export const getPermutiveAd = (adItem) => {
  const title = adItem.adtitle;
  const ad_system = adItem.adsystem;
  const duration = timeToSeconds(adItem.duration);
  return {
    duration,
    advertiser_name: '',
    ad_system,
    description: '',
    universal_ad_id_registry: '',
    creative_id: '',
    ad_id: '',
    min_suggested_duration: 30,
    title,
    universal_ad_id: '',
    deal_id: '',
  };
};

/**
 * Generate video object with given ad item for permutive tracking
 * @returns {object}
 */
export const getPermutiveVideo = (playlistItem) => {
  const video_id = playlistItem.mediaid;
  const name = playlistItem.title;
  const tags = playlistItem.tags ? playlistItem.tags.split(',') : [];
  const description = playlistItem.description;
  const duration = playlistItem.duration;
  const created_at = playlistItem.updated_at;
  const published_at = new Date(0); // The 0 there is the key, which sets the date to the epoch
  published_at.setUTCSeconds(playlistItem.pubdate);

  return {
    duration,
    name,
    custom_fields: [],
    video_id,
    description,
    tags,
    created_at,
    published_at,
  };
};

export const permutiveEventTracking = (eventName, trackingPayload) => {
  log(`permutiveEventTracking ${eventName}`);
  window.permutive.track(eventName, trackingPayload);
};

export const getCurrentPlaylistItem = (player) => {
  return player.getPlaylistItem(player.getPlaylistIndex());
};

/**
 * Set the initial query params to ad tag
 *
 * More details about the VAST ad tag URL parameters can be found on the document via this [link]{@link https://support.google.com/admanager/answer/10678356}
 * @param {import('./AdTag/AdTagURL').AdTagURL} adTag
 */
export const setInitialParamsToAdTag = (adTag) => {
  // it should be a hard code value
  adTag.searchParams.set(PARAM_SIZE, `640x480`);
  // it should be ad unit code instead of ad unit path
  adTag.searchParams.set(PARAM_AD_UNIT, JSGlobals.videoAdUnitPath);
  adTag.searchParams.set(PARAM_REQUEST_MODE, 's');
  adTag.searchParams.set(PARAM_DFP_REQUEST, '1');
  adTag.searchParams.set(PARAM_AD_RULE, '0');
  adTag.searchParams.set(PARAM_ENVIRONMENT, 'vp');
  adTag.searchParams.set(PARAM_AD_OUTPUT, 'vast');
  adTag.searchParams.set(PARAM_DELAYED_IMPRESSION, '1');
  adTag.searchParams.set(PARAM_DESCRIPTION_URL, MACRO_PAGE_URL);
  adTag.searchParams.set(PARAM_CORRELATOR, MACRO_TIMESTAMP);
  adTag.searchParams.set(PARAM_GDPR, MACRO_GDPR);
  // @see https://support.google.com/campaignmanager/answer/10031693?hl=en
  // `gdpr_consent=` must contain a valid TC string. If it does not, an ad may not be served or measured.
  if (JSGlobals.cmp.enabled) {
    adTag.searchParams.set(PARAM_GDPR_CONSENT, MACRO_GDPR_CONSENT);
    adTag.searchParams.set(PARAM_ADDTL_CONSENT, MACRO_ADDTL_CONSENT);
  }
  adTag.custParams.set('videoID', JSGlobals?.article?.heroMediaId);
  adTag.custParams.set('article', thisModule.getCustParamArticle());
  adTag.custParams.set('playerstate', 'hero');
  adTag.custParams.set('topictags', thisModule.getTopicTags());
  adTag.custParams.set('referrer', thisModule.getReferrer());
  adTag.custParams.set('mantis', thisModule.getMantisTargeting());
  adTag.custParams.set('pixels', thisModule.getPixelsTargeting());
  adTag.custParams.set('permutive', thisModule.getPermutive());
};

/**
 * @param {string} [url] - The URL to use for the ad tag.
 * @returns {import('./AdTag/AdTagURL').AdTagURL} adTag
 */
export const createAdTag = (url) => {
  const adTag = new AdTagURL(url);
  thisModule.setInitialParamsToAdTag(adTag);

  return adTag;
};

/**
 * Get the targeting key-value pairs from ad bidding services
 * @param {string} containerId
 * @param {string} mediaId
 * @param {string} adUnitCode
 * @param {string} adUnitPath
 * @returns {{[key: string], string}} targeting key-value pairs
 */
export const getBidsTargeting = async (
  containerId,
  mediaId,
  adUnitCode,
  _adUnitPath,
) => {
  await loadApsScript();

  const requests = [
    apsGetTargeting(adUnitCode),
    prebidGetTargeting(mediaId, containerId, adUnitCode, [640, 480]),
  ];

  const targetingArr = await Promise.allSettled(requests);
  const targeting = targetingArr.reduce((acc, curr) => {
    if (curr.status === 'fulfilled') {
      return { ...acc, ...curr.value };
    } else {
      warn(curr.reason);
    }
    return acc;
  }, {});

  return targeting;
};

/**
 * @typedef {object} TargetingOptions
 * @property {string} containerId
 * @property {string} mediaId
 * @property {string} adUnitCode
 * @property {string} adUnitPath
 */

/**
 * Initialize ad tag with the given parameters
 * @param {string} url
 * @param {TargetingOptions} options
 * @returns
 */
export const initAdTag = async (
  url,
  { containerId, mediaId, adUnitCode, adUnitPath },
) => {
  const adTag = thisModule.createAdTag(url);
  adTag.custParams.set('playerType', 'jwplayer');

  const isTV = thisModule.isTV();
  adTag.custParams.set('plcmt', isTV ? 1 : 2);

  const cookieDeprecationLabel = await getVideoCookieDeprecationLabel();
  if (cookieDeprecationLabel.length) {
    adTag.custParams.set('chrome_facilitated_testing', cookieDeprecationLabel);
  }

  const ppid = getPpid();
  if (ppid) {
    adTag.custParams.set('ppid', ppid);
  }

  const bidsTargeting = await getBidsTargeting(
    containerId,
    mediaId,
    adUnitCode,
    adUnitPath,
  );

  Object.entries(bidsTargeting).forEach(([key, value]) => {
    adTag.custParams.set(key, value);
  });

  return adTag.toString();
};
