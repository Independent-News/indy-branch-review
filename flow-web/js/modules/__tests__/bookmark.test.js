/**
 * @jest-environment jsdom
 */

import { CLASS_BOOKMARK_BUTTON } from '#app/constants/classNames';

import { InternalApi } from '#app/public/js/utils/internalApi';

import * as thatModule from '../bookmark';

jest.mock('#app/public/js/utils/internalApi');

describe('Bookmark public JS modules', () => {
  const mockButton = {
    classList: {
      contains: jest.fn(),
      add: jest.fn(),
      remove: jest.fn(),
    },
  };

  afterEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  describe('updateBookmarkState()', () => {
    const mockArgs = [null, mockButton];

    beforeEach(() => {
      jest.spyOn(thatModule, 'deleteBookmark').mockImplementation();
      jest.spyOn(thatModule, 'submitBookmark').mockImplementation();
    });

    it('will call deleteBookmark with passed args if button has saved class', () => {
      mockButton.classList.contains.mockReturnValueOnce(true);
      thatModule.updateBookmarkState(...mockArgs);
      expect(thatModule.deleteBookmark).toHaveBeenCalledTimes(1);
      expect(thatModule.deleteBookmark).toHaveBeenCalledWith(...mockArgs);
      expect(mockButton.classList.contains).toHaveBeenCalledTimes(1);
      expect(mockButton.classList.contains).toHaveBeenCalledWith('saved');
    });

    it('will call submitBookmark with passed args if button does not have saved class', () => {
      mockButton.classList.contains.mockReturnValueOnce(false);
      thatModule.updateBookmarkState(...mockArgs);
      expect(thatModule.submitBookmark).toHaveBeenCalledTimes(1);
      expect(thatModule.submitBookmark).toHaveBeenCalledWith(...mockArgs);
    });
  });

  describe('deleteBookmark()', () => {
    const mockArticleId = '12345';
    const mockArgs = [mockArticleId, mockButton];

    beforeEach(() => {
      document.body.innerHTML = `
      <button class="${CLASS_BOOKMARK_BUTTON} saved" data-article-id="${mockArticleId}"></button>
    `;
    });

    afterEach(() => {
      document.body.innerHTML = '';
    });

    it('will short circuit if button arg contains submitting class', () => {
      mockButton.classList.contains.mockReturnValueOnce(true);
      thatModule.deleteBookmark(...mockArgs);
      expect(mockButton.classList.contains).toHaveBeenCalledTimes(1);
      expect(mockButton.classList.contains).toHaveBeenCalledWith('submitting');
    });

    it('will make request to delete bookmark and update classes with local storage on successful request', async () => {
      mockButton.classList.contains.mockReturnValueOnce(false);
      InternalApi.delete.mockResolvedValueOnce({
        ok: true,
        json: jest.fn(() => Promise.resolve()),
      });
      jest.spyOn(thatModule, 'removeBookmarkLocalStorage').mockImplementation();
      await thatModule.deleteBookmark(...mockArgs);
      expect(mockButton.classList.add).toHaveBeenCalledTimes(1);
      expect(mockButton.classList.add).toHaveBeenCalledWith('submitting');
      expect(InternalApi.delete).toHaveBeenCalledTimes(1);
      expect(InternalApi.delete).toHaveBeenCalledWith(
        `bookmarks/${mockArticleId}`,
      );
      expect(mockButton.classList.remove).toHaveBeenCalledTimes(1);
      expect(mockButton.classList.remove).toHaveBeenCalledWith('submitting');
      expect(document.body.innerHTML).toMatchSnapshot();
      expect(thatModule.removeBookmarkLocalStorage).toHaveBeenCalledTimes(1);
      expect(thatModule.removeBookmarkLocalStorage).toHaveBeenCalledWith(
        mockArticleId,
      );
    });

    it('will maintain saved class and console error on unsuccessful delete request', async () => {
      jest.spyOn(console, 'error').mockImplementation();
      document.body.innerHTML = `
      <button class="${CLASS_BOOKMARK_BUTTON} saved" data-article-id="${mockArticleId}"></button>
    `;
      mockButton.classList.contains.mockReturnValueOnce(false);
      InternalApi.delete.mockResolvedValueOnce({
        ok: false,
      });
      await thatModule.deleteBookmark(...mockArgs);
      expect(mockButton.classList.remove).toHaveBeenCalledTimes(1);
      expect(mockButton.classList.remove).toHaveBeenCalledWith('submitting');
      expect(document.body.innerHTML).toMatchSnapshot();
      expect(console.error).toHaveBeenCalledTimes(1);
      expect(console.error).toHaveBeenCalledWith(
        'Fetch failed: ',
        'Network response was not ok.',
      );
    });
  });
});
