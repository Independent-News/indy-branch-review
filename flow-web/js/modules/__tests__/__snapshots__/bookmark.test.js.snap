// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark public JS modules deleteBookmark() will maintain saved class and console error on unsuccessful delete request 1`] = `
"
      <button class="bookmark-button saved" data-article-id="12345"></button>
    "
`;

exports[`Bookmark public JS modules deleteBookmark() will make request to delete bookmark and update classes with local storage on successful request 1`] = `
"
      <button class="bookmark-button" data-article-id="12345"></button>
    "
`;
